include/kos/wdt_timer.h

```diff
@@ -0,0 +1,106 @@
+    \ingroup kthreads
+
+    This file contains the watchdog timer API. A watchdog timer is a timer that
+    will trigger an action (through a pre-registered callback) after a timeout
+    expires. The code using a watchdog timer will typically reset it
+    periodically to prevent the action from happening.
+
+    \author Paul Cercueil
+
+    \see    kos/thread.h
+*/
+
+#ifndef __KOS_WDT_TIMER_H
+#define __KOS_WDT_TIMER_H
+
+#include <sys/cdefs.h>
+__BEGIN_DECLS
+
+struct wdt_timer;
+
+/** \struct  wdt_timer_t
+    \brief   Opaque structure describing one watchdog timer.
+*/
+typedef struct wdt_timer wdt_timer_t;
+
+/** \brief       Create a new watchdog timer.
+    \relatesalso wdt_timer_t
+
+    This function will create a watchdog timer using the specified callback,
+    programmed to expire after the given timeout. The timer will be stopped
+    by default and still need to be started using wdt_timer_start().
+
+    \param  cb              The function to call in case the watchdog timer
+                            expires.
+    \param  data            A parameter to pass to the function called.
+    \param  timeout_ms      The timeout value for the watchdog timer, in
+                            milliseconds.
+
+    \return                 The new watchdog timer on success, NULL on failure.
+
+    \sa wdt_timer_destroy, wdt_timer_start
+*/
+wdt_timer_t *wdt_timer_create(void (*cb)(void *), void *data,
+                              unsigned int timeout_ms);
+
+/** \brief       Stop and destroy a watchdog timer.
+    \relatesalso wdt_timer_t
+
+    This function will stop the watchdog timer and free its memory.
+
+    \param  timer           A pointer to the watchdog timer.
+
+    \sa wdt_timer_create
+*/
+void wdt_timer_destroy(wdt_timer_t *timer);
+
+/** \brief       Start a watchdog timer.
+    \relatesalso wdt_timer_t
+
+    This function will start the watchdog timer. If not stopped until the
+    timeout value expires, the watchdog timer will call the registered
+    callback function.
+
+    \param  timer           A pointer to the watchdog timer.
+
+    \sa wdt_timer_stop, wdt_timer_reset
+*/
+void wdt_timer_start(wdt_timer_t *timer);
+
+/** \brief       Stop a watchdog timer.
+    \relatesalso wdt_timer_t
+
+    This function will stop the watchdog timer. If it already expired, this
+    function does nothing. If it did not expire yet, the watchdog timer is
+    stopped and the registered callback won't be called.
+
+    \param  timer           A pointer to the watchdog timer.
+
+    \sa wdt_timer_start, wdt_timer_reset
+*/
+void wdt_timer_stop(wdt_timer_t *timer);
+
+/** \brief       Reset a watchdog timer.
+    \relatesalso wdt_timer_t
+
+    A convenience function to reset the watchdog timer by stopping it then
+    starting it again.
+
+    \param  timer           A pointer to the watchdog timer.
+
+    \sa wdt_timer_start, wdt_timer_stop
+*/
+static inline void wdt_timer_reset(wdt_timer_t *timer) {
```

Comment:
We suuuuuure we want this `static inline`? I mean, I get why we do, cuz it's a tiny wrapper, BUT that does mean that languages like D and Rust are not going to be able to bind to this, because there's no actual symbol being exported here... Is this needing to be high-performance?

Reply:
Languages like D and Rust can implement their own `timer.reset()` that calls two C functions instead of one...